runtime/compiler/env/CHTable.cpp

```diff
@@ -991,6 +1007,141 @@ bool CollectCompiledImplementors::visitSubclass(TR_PersistentClassInfo *cl)
+
+private:
+   std::vector<TR_OpaqueClassBlock *> _subClasses;
+   };
+
+bool
+CollectResolvedImplementors::visitSubclass(TR_PersistentClassInfo *cl)
+   {
+   TR_ASSERT_FATAL(comp()->isOutOfProcessCompilation(), "Must only be called on JITServer server-side");
+
+   auto callerMethod = static_cast<TR_ResolvedJ9JITServerMethod *>(_callerMethod);
+
+   // verify that our subclass meets all conditions
+   TR_OpaqueClassBlock *classId = cl->getClassId();
+   if (!TR::Compiler->cls.isAbstractClass(comp(), classId)
+      && !TR::Compiler->cls.isInterfaceClass(comp(), classId))
+      {
+      ++_numVisitedSubClasses;
+      if (_numVisitedSubClasses > _maxNumVisitedSubClasses)
+         {
+         // too many classes visited.
+         // set count greater than maxCount, to indicate failure and force exit
+         _count = _maxCount + 1;
+         stopTheWalk();
+         return false;
+         }
+
+      // if the implementor hasn't been cached yet,
+      // add the corresponding subclass to the list
+      TR_ResolvedMethod *resolvedMethod;
+      TR_ResolvedMethodType type = isInterface() ?
+         TR_ResolvedMethodType::Interface :
+         TR_ResolvedMethodType::VirtualFromOffset;
+      auto compInfoPT = static_cast<TR::CompilationInfoPerThreadRemote *>(static_cast<TR_J9VMBase *>(fe())->_compInfoPT);
+      TR_ResolvedMethodKey key =
+         compInfoPT->getResolvedMethodKey(
+            type,
+            reinterpret_cast<TR_OpaqueClassBlock *>(callerMethod->constantPoolHdr()),
+            _slotOrIndex,
+            classId);
+      if (!compInfoPT->getCachedResolvedMethod(
+             key,
+             callerMethod,
+             &resolvedMethod))
+         {
+         _subClasses.push_back(classId);
+         }
+      else if (resolvedMethod)
+         {
+         bool added = addImplementor(resolvedMethod);
+         if (added && _count >= _maxCount)
+            stopTheWalk();
+         }
+      else
+         {
+         // unresolved cached method found.
+         // set count greater than maxCount, to indicate failure and force exit
+         _count = _maxCount + 1;
+         stopTheWalk();
+         return false;
+         }
+      }
+
+   return true;
+   }
+
+void
+CollectResolvedImplementors::visitSubclasses(TR_PersistentClassInfo * classInfo, TR_CHTable::VisitTracker& visited)
+   {
+   TR_SubclassVisitor::visitSubclasses(classInfo, visited);
+   if (_depth == 0)
+      {
+      TR_ASSERT_FATAL(comp()->isOutOfProcessCompilation(), "Must only be called on JITServer server-side");
+      // _count >= _maxCount indicates that either all required implementors
+      // are already cached or that too many subclasses were visited.
+      // Remote call not needed.
+      if (_count >= _maxCount)
+         {
+         return;
+         }
+
+      auto callerMethod = static_cast<TR_ResolvedJ9JITServerMethod *>(_callerMethod);
+      callerMethod->cacheImplementorMethods(
```

Comment:
`visitSubclasses` is called from `visit` which acquires the CHTable mutex, thus we keep this mutex while sending a message. I would like to avoid that.

Reply:
Ok, I'll move this code back into a separate method then.